scipy/linalg/_decomp_svd.py

```diff
@@ -222,6 +249,21 @@ def svdvals(a, overwrite_a=False, check_finite=True):
     array([ 1.,  1.,  1.,  1.])
 
     """
+    xp = array_namespace(a)
+    if is_numpy(xp):
+        return _svdvals(a, overwrite_a=overwrite_a, check_finite=check_finite)
+    if overwrite_a:
+        raise ValueError(arg_err_msg("overwrite_a"))
+    if not check_finite:
+        raise ValueError(arg_err_msg("check_finite"))
+    if hasattr(xp, 'linalg'):
+        return xp.linalg.svdvals(a)
+    a = numpy.asarray(a)
+    return xp.asarray(_svdvals(a))
+
+
+def _svdvals(a, overwrite_a=False, check_finite=True):
+    # remove below line now we use array_namespace?
     a = _asarray_validated(a, check_finite=check_finite)
```

Comment:
can this `_asarray_validated` be removed now?

Reply:
Yes, but the `check_finite` behavior should be retained. Which means that you should be using `as_xparray` rather than `array_namespace` from `_lib/_array_api.py` I believe.

Reply:
Yes indeed. We should retain that behavior across all `scipy.linalg` functions.

Reply:
`_asarray_validated` includes one thing that `compliance_scipy` doesn't; a check for sparse matrices:
```python
if not sparse_ok:
    import scipy.sparse
    if scipy.sparse.issparse(a):
        msg = ('Sparse matrices are not supported by this function. '
               'Perhaps one of the scipy.sparse.linalg functions '
               'would work instead.')
        raise ValueError(msg)
```
Maybe we want to make a new utility which does just this sparse check, to avoid doing the rest of the checks twice?

Reply:
A sparse array will explicitly raise by default in `compliance_scipy`, since it's not convertible to a `numpy.ndarray`. So you only need it in places where `sparse_ok=True` is used. Having a separate check there sounds good.

Reply:
Ah, so `array_api_compat.array_namespace` will raise for the sparse arrays (I think this is what you mean, where does convertibility to an `ndarray` come into things?). Are we happy to just not support `scipy.sparse` arrays, while SCIPY_ARRAY_API is set (at least until `scipy.sparse` gets `__array_namespace__`)?

Reply:
Yes, it should raise, because the purpose of that function is to convert to an array API compatible/supported array type, and for unknown types to convert to `numpy.ndarray`. So what should happen is:
```python
>>> import numpy as np
>>> from scipy import sparse
>>> from scipy._lib._array_api import compliance_scipy
>>> a = sparse.csr_array((3, 4))
>>> compliance_scipy(a)
Traceback (most recent call last):
  Cell In[19], line 1
    compliance_scipy(a)
  File ~/code/scipy/build-install/lib/python3.10/site-packages/scipy/_lib/_array_api.py:46 in compliance_scipy
    raise TypeError("Only support Array API compatible arrays")
TypeError: Only support Array API compatible arrays
```

I don't know that it does though, because:
```python
>>> xp = array_namespace(a)
>>> xp.asarray(a)
array(<3x4 sparse array of type '<class 'numpy.float64'>'
        with 0 stored elements in Compressed Sparse Row format>, dtype=object)
```
so this may be a bug; conversion to `object` dtype is no longer supposed to happen with this machinery. Can you check?

Reply:
Resolving this now, the apparent bug above is not reproducible when the flag is set. This discussion around the checks in `compliance_scipy` has moved to https://github.com/scipy/scipy/issues/19118.